source/common/buffer/buffer_impl.cc

```diff
@@ -266,6 +266,30 @@ void* OwnedImpl::linearize(uint32_t size) {
   return slices_.front()->data();
 }
 
+RawSlice OwnedImpl::maybeLinearize(uint32_t max_size, uint32_t desired_min_size) {
```

Comment:
This implementation only look at the 1st and the 2nd slice in the buffer, would it makes more sense if we return linearized a group of whole slices, just before it exceed the max_size?

Reply:
Possibly. This is very much a heuristic, and there are a lot of ways this could potentially be improved. This improved performance of the case I was benchmarking, and didn't show any degradation for low-throughput (smaller request/response) traffic patterns, and was pretty simple and easy to reason about and/or predict what it will do.

Reply:
Yeah but I suggested the above because then we don't need the second parameter but keep pretty much same behavior for the rest.

Reply:
Can you elaborate on what you mean by "a linearized group of whole slices"? I don't understand what you're suggesting.

Reply:
like:
```
uint64_t size = 0;
for (slice : slices_) {
  if (size + slice_->dataSize() > max_size) return;
  size += slice_->dataSize();
}

return {linearize(size), size};
```

Reply:
Oh, I see, so in the case of a buffer of slices sized `{1, 16384, 16384, 16384...}`, with max_size 16384, this would yield (for successive calls) `{ 16385, 16384, 16384...}`. In terms of TLS records that will be produced, for that example I think it's roughly the same. Original version will be `{1, 16384, 16384, 16384...}` and your proposed version would produce `{16384, 1, 16384, 16384...}`, but it would copy an extra 16384 bytes (for the first call). But it would avoid the steady-state out-of-phase condition.

Reply:
No, my proposed version will do exactly same for `{1, 16384, 16384, 16384...}`, note it breaks before adding the slice that would exceed the size.

OTOH, if the incoming slices are `{1, 1, 16384, 16384, 16384...}`, the current version would do `{16384, 2, 16384, 16384...}` but mine will do `{2, 16384, 16384, 16384...}` without copying 16382 bytes.

Reply:
Oh, I see, that makes sense.

Another difference in behavior is if the slices are larger than `desired_min_size` and a series of them have a total size less than `max_size`. The original version would return each slice un-copied, and result in a TLS record for each. The new proposed version would combine them by copying, and only emit a single TLS record. It's not clear to me which one would be faster. Does anyone have an intuition on which is faster? @antoniovicente or @PiotrSikora?

Reply:
A single TLS record would be faster, that's the whole reason behind linearization.

Reply:
At some threshold the memcpy of linearization exceeds the overhead of the extra TLS record. For instance, if all slices were 16383 bytes (1 less than a full record), I don't believe it's faster to linearize everything to 16384; emitting slightly smaller records will be faster. 

I made a wild guess at picking 25% of a record as the threshold at which we should definitely linearize, and perf tests indicated that helped. But maybe I need to come up with a small benchmark to try to quantify this relationship.

Reply:
Right, but here we're comparing emitting many small TLS records vs combining many small buffers a single TLS record, not copying data between slices to completely fill records.

Reply:
In the example I was considering, I was asking about slices in the 4-8kb range. In that range it's not clear to me whether the memcpy cost will be more than the overhead of generating a TLS record.

Reply:
The difference between 2x 8KiB vs 1x 16KiB is probably negligible, but I'm pretty sure that `memcpy` overhead is smaller than writing additional TLS record(s) to the wire.

Reply:
...but that's an educated guess, feel free to benchmark this (e.g. by comparing proxy throughput, not userland microbenchmarks).